File: src/0061.rotate-list.61/solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function (head, k) {
    if (head === null || head.next === null) {
        return head;
    }
    // 先找出总数来
    let count = 0;
    let node = head;
    let prev;
    while (node !== null) {
        count++;
        prev = node;
        node = node.next;
    }
    // 处理rotate多轮的情况
    k = k % count;
    if (k === 0) {
        return head;
    }
    // 利用快慢指针找到倒数第k的节点
    let slow = head;
    let fast = head;
    while (k > 0) {
        k--;
        fast = fast.next;
    }
    while (fast.next !== null) {
        slow = slow.next;
        fast = fast.next;
    }
    // rotate
    node = slow.next;
    slow.next = null;
    prev.next = head;

    return node;
};

File: src/0061.rotate-list.61/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(head == null){
            return null;
        }
        int L = getLength(head);
        k %= L;
        if(k == 0){
            return head;
        }
        ListNode slow = head;
        ListNode fast = head;
        while (k>0){
            fast = fast.next;
            k--;
        }
        while (fast.next != null){
            slow = slow.next;
            fast = fast.next;
        }
        ListNode next = slow.next;
        slow.next = null;
        fast.next = head;
        
        

        return next;
    }

    private int getLength(ListNode head){
        int count = 0;
        while (head != null){
            head = head.next;
            count++;
        }
        return count;
    }
}